### PR TITLE
Change JComponentHelper filter usage to namespace version

### PR DESCRIPTION
**Pull request by HLeithner**
Pull Request for Issue #41164 .

### Summary of Changes
This commit changes all usages of JComponentHelper to \Joomla\CMS\Component\ComponentHelper. Additionally an exception is thrown when JComponehtHelper is used as filter but is not resolvable.

@richard67 we need a migration script for the database entries... at least for the plugins and the custom fields. Any idea what's the best way is for this?

### Testing Instructions
Open Article add html save it.

### Actual result BEFORE applying this Pull Request
HTML format get removed

### Expected result AFTER applying this Pull Request
HTML format still here

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed

- [X] Pull Request link for manual.joomla.org: <link> Needs documentation that JComponentHelper is not resolved anymore
- [ ] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by richard67** [comment]
@HLeithner A migration function could be called in the "postflight" method in script.php similar to here:

https://github.com/joomla/joomla-cms/blob/5.0-dev/administrator/components/com_admin/script.php#L2202-L2205

The method function be a new private function similar to the one called in the example above https://github.com/joomla/joomla-cms/blob/5.0-dev/administrator/components/com_admin/script.php#L2217 .

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on d0a2509484366bf4439b8266c1debfe3e88d1e7f<br /><br />tested the editor with the backward compatibility plugin disabled and the formatting was kept<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/41259">issues.joomla.org/tracker/joomla-cms/41259</a>.</sub>

**Comment by HLeithner** [comment]
> @HLeithner A migration function could be called in the "postflight" method in script.php similar to here:
> 
> https://github.com/joomla/joomla-cms/blob/5.0-dev/administrator/components/com_admin/script.php#L2202-L2205
> 
> It might need to modify the version check before that so it runs also when updating from a 5.0.0-alpha, i.e. anything lower than 5.0.0-beta1: https://github.com/joomla/joomla-cms/blob/5.0-dev/administrator/components/com_admin/script.php#L2198-L2200
> 
> The method function be a new private function similar to the one called in the example above https://github.com/joomla/joomla-cms/blob/5.0-dev/administrator/components/com_admin/script.php#L2217

can you help me writing this function please?

**Comment by Fedik** [comment]
I would suggest to make a filter class, instead.
Example like for `filter="safehtml"` https://github.com/joomla/joomla-cms/blob/e2170c8a6224ebb6ed754853de0d5e1b97ba1ce9/libraries/src/Form/Filter/SafehtmlFilter.php#L43-L51

That will call `Joomla\CMS\Component\ComponentHelper::filterText` internally:
```php
public function filter($element, $value ...) {
  return Joomla\CMS\Component\ComponentHelper::filterText($value);
}
```

**Comment by wilsonge** [comment]
I actually agree with Fedik here. I think given this is a very common pattern in core making lots of people hardcode php callbacks is asking for b/c breaks (because extensions will be using this too). Making it a static name will be more maintainable going forwards.

**Comment by HLeithner** [comment]
have you looked at the `loadType` how complicate it is to do the same thing then just setting the real name of the filter class+function.

It also uses J-prefix for classes, I'm not even sure which namespaces it considers, it might look only for Joomla core namespace? maybe I'm wrong. But this construct to just load one of our 5 filters is completely over engineered, if we allow such thing like shortcut to a filter object method then I think hardcoding that 5 classes makes more sense then expecting any 3rd party dev to reverse engineer this function...

never the less I'm happy for a shortcut for simple text filterText, but I would deprecate/remove this class autoloading voodoo

**Comment by richard67** [comment]
> > @HLeithner A migration function could be called in the "postflight" method in script.php similar to here:
> > https://github.com/joomla/joomla-cms/blob/5.0-dev/administrator/components/com_admin/script.php#L2202-L2205
> > It might need to modify the version check before that so it runs also when updating from a 5.0.0-alpha, i.e. anything lower than 5.0.0-beta1: https://github.com/joomla/joomla-cms/blob/5.0-dev/administrator/components/com_admin/script.php#L2198-L2200
> > The method function be a new private function similar to the one called in the example above https://github.com/joomla/joomla-cms/blob/5.0-dev/administrator/components/com_admin/script.php#L2217
> 
> can you help me writing this function please?

@HLeithner Don‘t expect any help or advise from me because when I give some you obviously ignore it.

**Comment by HLeithner** [comment]
ok no problem thanks

**Comment by bayareajenn** [comment]
In reading through the comments, would you like me to test this now or wait until more changes are made?

**Comment by HLeithner** [comment]
Yes please test it (with new installatons for now only, updates without b/c plugin are not handled yet)

**Comment by brianteeman** [comment]
@bayareajenn the code works - the question is if it is an appropriate solution

**Comment by bayareajenn** [comment]
I have tested this item :white_check_mark: successfully on d0a2509484366bf4439b8266c1debfe3e88d1e7f<br /><br />Tested with b/c plugin enabled and disabled in articles, categories, and custom modules. Works in all cases. Thanks :)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/41259">issues.joomla.org/tracker/joomla-cms/41259</a>.</sub>

**Comment by richard67** [comment]
> > @HLeithner A migration function could be called in the "postflight" method in script.php similar to here:
> > https://github.com/joomla/joomla-cms/blob/5.0-dev/administrator/components/com_admin/script.php#L2202-L2205
> > It might need to modify the version check before that so it runs also when updating from a 5.0.0-alpha, i.e. anything lower than 5.0.0-beta1: https://github.com/joomla/joomla-cms/blob/5.0-dev/administrator/components/com_admin/script.php#L2198-L2200
> > The method function be a new private function similar to the one called in the example above https://github.com/joomla/joomla-cms/blob/5.0-dev/administrator/components/com_admin/script.php#L2217
> 
> can you help me writing this function please?

@HLeithner What needs to be migrated?

**Comment by HLeithner** [comment]
yes, at least com_fields needs to be updated JComponentHelper::filterText -> to the namespaced version

**Comment by richard67** [comment]
> yes, at least com_fields needs to be updated JComponentHelper::filterText -> to the namespaced version

Will see if I can provide something tomorrow.

**Comment by HLeithner** [comment]
thanks

**Comment by richard67** [comment]
@HLeithner As the replacement would be a complete parameter value, it could be done easily with an update SQL script and string replacement. But that is not really nice with json strings, even if it would be safe, so the other way would be a conversion method in script.php like I suggested above. What do you prefer? Easy but theoretical risky way in SQL? Or safer but maybe a bit over engineered way in PHP?

**Comment by HLeithner** [comment]
Thats a good question... We could do it with SQL json functions...?

I'm happy with both

**Comment by richard67** [comment]
> Thats a good question... We could do it with SQL json functions...?
> 
> I'm happy with both

I'm not familiar with json functions in SQL.

**Comment by richard67** [comment]
@HLeithner I've just made a new installation on a MySQL 8 database with the branch of this PR and then immediately after the installation I've checked the parameters in database. They lose the backslash, so for the editor field they are e.g.:
`{"buttons":0,"width":"100%","height":"250px","filter":"JoomlaCMSComponentComponentHelper::filterText"}`

The form shows "None" as selected filter for that field.

If I change the selection in the form to "Text" and save the form, the result in database is with double backslashes:
`{"buttons":0,"width":"100%","height":"250px","filter":"\\Joomla\\CMS\\Component\\ComponentHelper::filterText"}`

**So we have to use double backslashes in the base.sql scripts** for the `params`'s `filter` values:
````
(0, 'plg_fields_editor', 'plugin', 'editor', 'fields', 0, 1, 1, 0, 1, '', '{"buttons":0,"width":"100%","height":"250px","filter":"\\Joomla\\CMS\\Component\\ComponentHelper::filterText"}', '', 4, 0),
````

I haven't testes PostgreSQL yet, but I expect it to be the same.

The XML files are ok, no need for double backslashes there.

I've also experimented with an update SQL statement for an update script. On MySQL the following works with double backslashes:

```
UPDATE `#__extensions`
   SET `params` = JSON_REPLACE(`params`, '$.filter' , '\\Joomla\\CMS\\Component\\ComponentHelper::filterText')
 WHERE `type` = 'plugin'
   AND `folder` = 'fields'
   AND `params` <> ''
   AND JSON_EXTRACT(`params`, '$.filter') = 'JComponentHelper::filterText';
```
This would change it for all field plugins which have a filter property equal to "JComponentHelper::filterText" in their params.

Should we limit this to the editor and textarea plugins with an additional `AND element in ('editor', 'textarea')`?

If that also works for MariaDB and there is something similar for PostgreSQL, we can use an update SQL script.

File: installation/sql/mysql/base.sql [hunk 1]
```diff
@@ -299,7 +299,7 @@ INSERT INTO `#__extensions` (`package_id`, `name`, `type`, `element`, `folder`,
 (0, 'plg_fields_sql', 'plugin', 'sql', 'fields', 0, 1, 1, 0, 1, '', '', '', 10, 0),
 (0, 'plg_fields_subform', 'plugin', 'subform', 'fields', 0, 1, 1, 0, 1, '', '', '', 11, 0),
 (0, 'plg_fields_text', 'plugin', 'text', 'fields', 0, 1, 1, 0, 1, '', '', '', 12, 0),
-(0, 'plg_fields_textarea', 'plugin', 'textarea', 'fields', 0, 1, 1, 0, 1, '', '{"rows":10,"cols":10,"maxlength":"","filter":"JComponentHelper::filterText"}', '', 13, 0),
+(0, 'plg_fields_textarea', 'plugin', 'textarea', 'fields', 0, 1, 1, 0, 1, '', '{"rows":10,"cols":10,"maxlength":"","filter":"\Joomla\CMS\Component\ComponentHelper::filterText"}', '', 13, 0),
```
Comment (richard67) [inline installation/sql/mysql/base.sql]:
```suggestion
(0, 'plg_fields_textarea', 'plugin', 'textarea', 'fields', 0, 1, 1, 0, 1, '', '{"rows":10,"cols":10,"maxlength":"","filter":"\\Joomla\\CMS\\Component\\ComponentHelper::filterText"}', '', 13, 0),
```
We have to use double backslashes, see my comment https://github.com/joomla/joomla-cms/pull/41259#issuecomment-1686014933 .

File: installation/sql/mysql/base.sql [hunk 2]
```diff
@@ -290,7 +290,7 @@ INSERT INTO `#__extensions` (`package_id`, `name`, `type`, `element`, `folder`,
 (0, 'plg_fields_calendar', 'plugin', 'calendar', 'fields', 0, 1, 1, 0, 1, '', '', '', 1, 0),
 (0, 'plg_fields_checkboxes', 'plugin', 'checkboxes', 'fields', 0, 1, 1, 0, 1, '', '', '', 2, 0),
 (0, 'plg_fields_color', 'plugin', 'color', 'fields', 0, 1, 1, 0, 1, '', '', '', 3, 0),
-(0, 'plg_fields_editor', 'plugin', 'editor', 'fields', 0, 1, 1, 0, 1, '', '{"buttons":0,"width":"100%","height":"250px","filter":"JComponentHelper::filterText"}', '', 4, 0),
+(0, 'plg_fields_editor', 'plugin', 'editor', 'fields', 0, 1, 1, 0, 1, '', '{"buttons":0,"width":"100%","height":"250px","filter":"\Joomla\CMS\Component\ComponentHelper::filterText"}', '', 4, 0),
```
Comment (richard67) [inline installation/sql/mysql/base.sql]:
```suggestion
(0, 'plg_fields_editor', 'plugin', 'editor', 'fields', 0, 1, 1, 0, 1, '', '{"buttons":0,"width":"100%","height":"250px","filter":"\\Joomla\\CMS\\Component\\ComponentHelper::filterText"}', '', 4, 0),
```
We have to use double backslashes, see my comment https://github.com/joomla/joomla-cms/pull/41259#issuecomment-1686014933 .

File: installation/sql/postgresql/base.sql [hunk 3]
```diff
@@ -296,7 +296,7 @@ INSERT INTO "#__extensions" ("package_id", "name", "type", "element", "folder",
 (0, 'plg_fields_calendar', 'plugin', 'calendar', 'fields', 0, 1, 1, 0, 1, '', '', '', 1, 0),
 (0, 'plg_fields_checkboxes', 'plugin', 'checkboxes', 'fields', 0, 1, 1, 0, 1, '', '', '', 2, 0),
 (0, 'plg_fields_color', 'plugin', 'color', 'fields', 0, 1, 1, 0, 1, '', '', '', 3, 0),
-(0, 'plg_fields_editor', 'plugin', 'editor', 'fields', 0, 1, 1, 0, 1, '', '{"buttons":0,"width":"100%","height":"250px","filter":"JComponentHelper::filterText"}', '', 4, 0),
+(0, 'plg_fields_editor', 'plugin', 'editor', 'fields', 0, 1, 1, 0, 1, '', '{"buttons":0,"width":"100%","height":"250px","filter":"\Joomla\CMS\Component\ComponentHelper::filterText"}', '', 4, 0),
```
Comment (richard67) [inline installation/sql/postgresql/base.sql]:
```suggestion
(0, 'plg_fields_editor', 'plugin', 'editor', 'fields', 0, 1, 1, 0, 1, '', '{"buttons":0,"width":"100%","height":"250px","filter":"\\Joomla\\CMS\\Component\\ComponentHelper::filterText"}', '', 4, 0),
```
We have to use double backslashes, see my comment https://github.com/joomla/joomla-cms/pull/41259#issuecomment-1686014933 .

File: installation/sql/postgresql/base.sql [hunk 4]
```diff
@@ -305,7 +305,7 @@ INSERT INTO "#__extensions" ("package_id", "name", "type", "element", "folder",
 (0, 'plg_fields_sql', 'plugin', 'sql', 'fields', 0, 1, 1, 0, 1, '', '', '', 10, 0),
 (0, 'plg_fields_subform', 'plugin', 'subform', 'fields', 0, 1, 1, 0, 1, '', '', '', 11, 0),
 (0, 'plg_fields_text', 'plugin', 'text', 'fields', 0, 1, 1, 0, 1, '', '', '', 12, 0),
-(0, 'plg_fields_textarea', 'plugin', 'textarea', 'fields', 0, 1, 1, 0, 1, '', '{"rows":10,"cols":10,"maxlength":"","filter":"JComponentHelper::filterText"}', '', 13, 0),
+(0, 'plg_fields_textarea', 'plugin', 'textarea', 'fields', 0, 1, 1, 0, 1, '', '{"rows":10,"cols":10,"maxlength":"","filter":"\Joomla\CMS\Component\ComponentHelper::filterText"}', '', 13, 0),
```
Comment (richard67) [inline installation/sql/postgresql/base.sql]:
```suggestion
(0, 'plg_fields_textarea', 'plugin', 'textarea', 'fields', 0, 1, 1, 0, 1, '', '{"rows":10,"cols":10,"maxlength":"","filter":"\\Joomla\\CMS\\Component\\ComponentHelper::filterText"}', '', 13, 0),
```
We have to use double backslashes, see my comment https://github.com/joomla/joomla-cms/pull/41259#issuecomment-1686014933 .

**Comment by richard67** [comment]
@HLeithner I've created PR #41417 with my changes suggested above for the base.sql scripts and an update SQL for migrating the parameter.

The update SQL currently would change it for all field plugins which have a filter property equal to "JComponentHelper::filterText" in their params, which would also include 3rd party field plugins.

Should we limit this to the editor and textarea plugins with an additional `AND element in ('editor', 'textarea')` so that it's limited to the core only?

**Comment by HLeithner** [comment]
I wouldn't limit it, because this would also convert 3rd party fields... Thinking about this might be a problem... Maybe you are right and we limit it to (relevant) core fields only

**Comment by richard67** [comment]
> I wouldn't limit it, because this would also convert 3rd party fields... Thinking about this might be a problem... Maybe you are right and we limit it to (relevant) core fields only

@HLeithner Relevant core fields plugins are only 'editor', 'text' and 'textarea', is that right?

**Comment by richard67** [comment]
> > I wouldn't limit it, because this would also convert 3rd party fields... Thinking about this might be a problem... Maybe you are right and we limit it to (relevant) core fields only
> 
> @HLeithner Relevant core fields plugins are only 'editor', 'text' and 'textarea', is that right?

I've updated my PR to limit the migration to the 3 named plugins.

**Comment by richard67** [comment]
@HLeithner You wrote in the description of this PR:

> @richard67 we need a migration script for the database entries... at least for the plugins and the custom fields. Any idea what's the best way is for this?

For the plugins we have my PR for you.

But what needs to be done for custom fields?

**Comment by richard67** [comment]
I've updated my PR #41417 by the migration of the custom field's fieldparams on update. I haven't tested it yet, but it should work.

**Comment by laoneo** [comment]
Merging this for now to have the description fields accepting HTML content again. Thanks @richard67 for your patience on this.

This pr doesn't prevent to add a more sophisticated solution in the future, as @Fedik suggested.

**Pull request merged**